File: LeetCode/88MergeSortedArray/MergeSortedArray.cpp
class Solution {
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {

        // Without temp array -> Time: O(n), Space: O(1)
        while ( (m > 0) && (n > 0)) {
            if (nums1[m - 1] > nums2[n - 1]) {
                nums1[m+n-1] = nums1[m - 1];
                m--;
            }
            else {
                nums1[m+n-1] = nums2[n - 1];
                n--;
            }
        }
        while (n > 0) {
            nums1[m+n-1] = nums2[n - 1];
            n--;
        }

        // // With temp array -> Time: O(n), Space: O(n)
        // vector<int> nums3(m+n);
        // int i = 0;
        // int j = 0;
        // while ( (i < m) && (j < n)) {
        //     if (nums1[i] < nums2[j]) {
        //         nums3[i+j] = nums1[i];
        //         i++;
        //     }
        //     else {
        //         nums3[i+j] = nums2[j];
        //         j++;
        //     }
        // }
        // while (i < m) {
        //     nums3[i+n] = nums1[i];
        //     i++;
        // }
        // while (j < n) {
        //     nums3[j+m] = nums2[j];
        //     j++;
        // }
        // for (int k = 0; k < m+n; k++) {
        //     nums1[k] = nums3[k];
        // }
    }
};

File: LeetCode/88MergeSortedArray/MergeSortedArray.py
class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """
        # Time: O(n), Space: O(1)
        i, j = m-1, n-1
        while i > -1 and j > -1:
            if nums1[i] > nums2[j]:
                nums1[i+j+1] = nums1[i]
                i -= 1
            else:
                nums1[i+j+1] = nums2[j]
                j -= 1
        while i > -1:
            nums1[i+j+1] = nums1[i]
            i -= 1
        while j > -1:
            nums1[i+j+1] = nums2[j]
            j -= 1
        return nums1

        # # Without temp array -> Time: O(n), Space: O(1)
        # last = m + n - 1
        # while m > 0 and n > 0:
        #     if nums1[m - 1] > nums2[n - 1]:
        #         nums1[last] = nums1[m - 1]
        #         m -= 1
        #     else:
        #         nums1[last] = nums2[n - 1]
        #         n -= 1
        #     last -= 1
        # while n > 0:
        #     nums1[last] = nums2[n - 1]
        #     n -= 1
        #     last -= 1


        # # With temp array -> Time: O(n), Space: O(n)
        # nums3 = []
        # i = 0
        # j = 0
        # while i < m and j < n:
        #     if nums1[i] < nums2[j]:
        #         nums3.append(nums1[i])
        #         i += 1
        #     else:
        #         nums3.append(nums2[j])
        #         j += 1
        # while i < m:
        #     nums3.append(nums1[i])
        #     i += 1
        # while j < n:
        #     nums3.append(nums2[j])
        #     j += 1
        # for i in range(len(nums3)):
        #     nums1[i] = nums3[i]
